### PR TITLE
Training documentation

### PR DESCRIPTION
**Pull request by LorenzLamm**
This PR contains mainly documentation, but also the "patch_correction" module that can be used to easily extract patches from the tomogram that can be used for further training.

Main additions:
1. Documentation about training a new model
2. Documentation about how to create new annotations for training a new model
3. "patch_corrections" module, containing functionalities to extract patches & merge corrections into a training patch

### CODE REVIEW

File: README.md
Comment (LorenzLamm) [inline README.md:1]: I put some more information to the readme file to have it accessible directly from the start page (One link less to click is always good I guess)

Comment (LorenzLamm) [reply README.md:1]: Also added the link to a pretrained model

File: docs/Usage/Annotations.md
Comment (LorenzLamm) [inline docs/Usage/Annotations.md:1]: This file describes our strategy to create and annotate new training patches, and also how to use our new scripts to extract the patches & merge the corrections.

Comment (LorenzLamm) [reply docs/Usage/Annotations.md:1]: (btw our student Simon also created some YouTube tutorials for this, which I linked also here)

File: docs/Usage/Segmentation.md
Comment (LorenzLamm) [inline docs/Usage/Segmentation.md:1]: I added the link to the pretrained model & a screenshot that shows the CLI for the segmentation.
I guess this is the most important functionality of the package, so I hope the documentation is clear as it is.

File: docs/Usage/Training.md
Comment (LorenzLamm) [inline docs/Usage/Training.md:1]: Also added some instructions on how to perform your own training here.

File: docs/installation.md
Comment (LorenzLamm) [inline docs/installation.md:1]: Without the editable installation, imports of the packages do not work :/
As a workaround, I added the -e flag again, but would be cool if we could solve this in a follow-up

File: src/annotations/extract_patch_cli.py
Comment (LorenzLamm) [inline src/annotations/extract_patch_cli.py:1]: The extract_patches function uses either x,y,z-coordinates as direct input, or a .csv file with multiple coordinates.
Then it extracts patches from the tomogram (and segmentation) centered around the given coordinates.

File: src/annotations/merge_corrections.py
Comment (LorenzLamm) [inline src/annotations/merge_corrections.py:1]: The "merge_corrections" function reads the manual corrections done on each patch, and merges them together with the original segmentation.

File: src/membrain_seg/cli/cli.py
Comment (LorenzLamm) [inline src/membrain_seg/cli/cli.py:1]: Added the "rich" markup mode to make it look nicer :)

File: src/membrain_seg/cli/segment_cli.py
Comment (LorenzLamm) [inline src/membrain_seg/cli/segment_cli.py:1]: The parameters are described also further below in the "help" argument of the typer CLI, so I thought it's more clean to remove it here, and instead use an example on how to use the function.

File: README.md
Comment (kevinyamauchi) [reply README.md:1]: Cool. I agree it's nice to make it easier for users to find the information they need!

File: docs/Usage/Annotations.md
Comment (kevinyamauchi) [reply docs/Usage/Annotations.md:1]: Also, this is super nice. Looking forward to making this into a napari plugin ;)

File: docs/Usage/Segmentation.md [hunk 1]
```diff
@@ -22,12 +22,24 @@ For the prediction, you will basically need two files:
 It may make sense to use a preprocessed tomogram.
 2. A **pre-trained MemBrain segmentation model**
 
-For point 2, as we are still in the development phase, we do not have a public model right now.
-If you would like to give MemBrain-seg a try, we are happy to provide our latest model.
-Just write an email to lorenz.lamm@helmholtz-munich.de
+A pre-trained MemBrain segmentation model can be downloaded [here](https://drive.google.com/file/d/15ZL5Ao7EnPwMHa8yq5CIkanuNyENrDeK/view?usp=sharing).
```
Comment (kevinyamauchi) [inline docs/Usage/Segmentation.md:25]: This is fine for now, but I think we should put it on zenodo relatively soon. That way the model is versioned. Also, we can automatically download it with `pooch`.

File: docs/installation.md
Comment (kevinyamauchi) [reply docs/installation.md:1]: yes, let's fix this in a follow up.

File: src/annotations/extract_patch_cli.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    --segmentation-path <path-to-your-segmentation> --out-folder <path-to-out-folder>
+    --x 100 --y 200 --z 300
+
+    """
+    if coords_file is not None:
+        coords = get_csv_data(coords_file)
+    else:
+        coords = [np.array((x, y, z))]
+    _extract_patches(
+        tomogram_path, segmentation_path, coords, out_folder, idx_add, token
```
Comment (kevinyamauchi) [inline src/annotations/extract_patch_cli.py:87]: Not a blocker, but I like to use keyword arguments instead of positional arguments. that way if the function argument order or names change, it breaks rather than running with potentially the wrong parameterization.

File: src/annotations/merge_corrections.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        for filename in os.listdir(corrections_dir):
+            if not os.path.isdir(os.path.join(corrections_dir, filename)):
+                continue
+            if filename == token:
+                cur_patch_corrections_folder = os.path.join(corrections_dir, filename)
+                merged_corrections = get_corrections_from_folder(
+                    cur_patch_corrections_folder, os.path.join(labels_dir, label_file)
+                )
+                out_file = os.path.join(out_dir, label_file)
+                print("Storing corrections in", out_file)
```
Comment (kevinyamauchi) [inline src/annotations/merge_corrections.py:119]: not a blocker, but it would be nice to use the warnings module instead of print so that these can be easily supressed if desired. I think it's fine to do this in a follow-up PR.,

**Review by kevinyamauchi — APPROVED** [review]
Looks good to me! I left some minor comments below. Please feel free to merge once you've taken a look at them.

File: docs/Usage/Training.md [hunk 4]
```diff
@@ -1,3 +1,69 @@
 # Training
+MemBrain-seg is designed to work out-of-the-box and ideally will not require training your own model.
```
[the remark this answers is not in the document]
Comment (LorenzLamm) [reply docs/Usage/Training.md:2]: Absolutely! I haven't tested what parameters would work best for fine-tuning (probably wouldn't need to change much, maybe start with a smaller learning rate than usual).
It's definitely on my to-do-list!

File: src/annotations/merge_corrections.py (hunk 3 above)
Comment (LorenzLamm) [reply src/annotations/merge_corrections.py:119]: Ah yes! Will do that in a follow-up PR :)

File: src/annotations/extract_patch_cli.py (hunk 2 above)
Comment (LorenzLamm) [reply src/annotations/extract_patch_cli.py]: I changed to keyword arguments for this and some other functions, but should check for the whole repository in a follow-up and see where I can add more.

File: src/annotations/extract_patches.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+)
+
+
+class InvalidCoordinatesError(Exception):
+    """Exception raised of coordinates exceed tomogram boundaries."""
+
+    pass
+
+
+def pad_labels(patch, padding):
```
[the remark this answers is not in the document]
Comment (LorenzLamm) [reply src/annotations/extract_patches.py]: Agreed! Added this as a parameter to the function.

Currently, training assumes 2.0 to be the ignore label, but would be cool to make that also flexible in a follow-up.
This way, training could potentially also be used for other classes, and not only membranes.

File: src/membrain_seg/dataloading/data_utils.py [hunk 6]
```diff
@@ -23,6 +23,67 @@ def make_directory_if_not_exists(path: str):
...
+    Exception
+        If the data can't be converted to float numpy array, a numpy array with
+        original type will be returned.
+
+    """
+    rows = []
+    with open(csv_path) as csv_file:
+        csv_reader = csv.reader(csv_file, delimiter=delimiter)
+        for row in csv_reader:
+            rows.append(row)
```
[the remark this answers is not in the document]
Comment (LorenzLamm) [reply src/membrain_seg/dataloading/data_utils.py:65]: I put it on my to-do-list. There is also some other csv reading in the package. I'll try to make them consistent
